### ISSUE TITLE
URL review

### ISSUE THREAD

**Issue by paulirish** [GoogleChrome/lighthouse-ci#40]
Could we do a once-over all on the webserver URLs? 
And also review the rest API?

I remember thinking that the compare view should have 'compare' in it, etc. But I figure we should do this sooner than later since it's very breaking changey.

@patrickhulce  what's the best way to review all these?

**Comment by patrickhulce** [comment]
webserver URLs 
- /app/projects
- /app/projects/:projectId
- /app/projects/:projectId/builds/:buildId

https://github.com/GoogleChrome/lighthouse-ci/blob/0c1c082f26abb93e0701f9fea3362434bb76a35a/packages/server/src/ui/app.jsx#L21-L43

rest API URLs

- /v1/projects
- /v1/projects/lookup
- /v1/projects/${projectId}/token
- /v1/projects/${projectId}/builds
- /v1/projects/${projectId}/builds
- /v1/projects/${projectId}/branches
- /v1/projects/${projectId}/urls
- /v1/projects/${projectId}/builds/${buildId}/ancestor
- /v1/projects/${projectId}/builds/${buildId}/runs
- /v1/projects/${projectId}/builds/${buildId}/urls
- /v1/projects/${projectId}/builds/${buildId}/statistics
- /v1/projects/${projectId}/builds/${buildId}/runs

https://github.com/GoogleChrome/lighthouse-ci/blob/0c1c082f26abb93e0701f9fea3362434bb76a35a/packages/utils/src/api-client.js#L137-L261

**Comment by patrickhulce** [comment]
> what's the best way to review all these?

perhaps just discussion on this issue now that there's a list and links to the source :)

**Comment by paulirish** [comment]
> /app/projects

makes sense

> /app/projects/:projectId

kinda wish there was a friendlier URL instead of the super long UUID.. but i can imagine this adds a good deal of complexity and we can always improve on this later.

> /app/projects/:projectId/builds/:buildId

1) i think this should be `/compare/`. If hypothetically we end up showing something like [pinpoint](https://pinpoint-dot-chromeperf.appspot.com/job/15b70c30220000).... oops. no, they use `/job/` for a ... comparison view.. Well there goes my defense. :)  

Interestingly when you do baseHash then you have both a git hash and a buildId in the url (the latter isn't a git hash but looks close to one)

and to the user the git hash seems like an almost canonical representation of that page.  So i feel like the user expects us to do `/app/:friendlyProjectName/compare/:compareGitHash`

... But... i guess I see why we're being conservative here.

**Comment by paulirish** [comment]
I was ready to close this ticket but had one more thought.

How about using something shorter than a full UUID in both cases. Like we can see [pinpoint is using](https://pinpoint-dot-chromeperf.appspot.com/job/15b70c30220000) some 14 character hexadecimal string.

I'd prefer something a bit more succinct.

how about something like...

```
/app/projects/:projectSlug_:projectIdShortUUID/build/:buildIdShortUUID
```

we could be like amazon with their readable URLs that still contain the product ID: `https://www.amazon.com/Koss-UR10-Headphones-Durable-3-5mm/dp/B0001IM95U/`

**Comment by paulirish** [comment]
also... /builds/ or /build/  ? (and same for projects..)

github uses `/issues/7329` but then they also use `/pull/3831`. lol.

i personally prefer singular noun for viewing a single thing.

**Comment by connorjclark** [comment]
> i personally prefer singular noun for viewing a single thing.

if you have a box of hats in your closet, what label would you put on the box?

**Comment by exterkamp** [comment]
> if you have a box of hats in your closet, what label would you put on the box?

old hat(s)

**Comment by paulirish** [comment]
There is only one hat in this box.

![image](https://user-images.githubusercontent.com/39191/68356131-5bafe480-00c6-11ea-8b68-703c376b5999.png)

**Comment by paulirish** [comment]
(to clarify i'm talking about the web URLs not the api. i agree the API should stay plural.)

**Comment by patrickhulce** [comment]
tl;dr - how about `/app/:friendlyProjectSlug/compare/:compareGitHash`?


Mind explaining the motivation behind shorter URLs? :)

> How about using something shorter than a full UUID in both cases

Going with a bespoke ID is fine, I'd just like to have a compelling reason to move from an industry standard to something we come up with and right now it seems like the human readability of the URL is the only reason we have. I'm not sure how much more useful a link with two sets of 14 random characters is than a link with two sets of 36 random characters :)

> /app/:friendlyProjectName/compare/:compareGitHash

Replacing the build ID with hash is totally 👍 to me, since we're already enforcing 1-1 build Id <-> hash relationship

**Comment by paulirish** [comment]
> Mind explaining the motivation behind shorter URLs? :)

sure! yes this is entirely about human readability. computers can read much better, so we can afford to make it harder for them if it means a UX boon for people.

you know when you copypaste a url to someone and it has a million queryparams on it and so it fulls the whole hangouts window and you're like "OH SORRY!"

or when you get an email that has a link in that's both tempting and super-suspicious because it's incomprehensible:
![image](https://user-images.githubusercontent.com/39191/68363516-ef41df00-00df-11ea-8720-1c002d9124fd.png)

I just really appreciate well designed URLs. The best ones provide the perfect context to the user who's about to click on one.  :)

When frontend developer Cassandra writes that PR that improves the company webapp's perf score by 10 points, she's going to take that compare link and share it every which way. I want people to feel comfortable about clicking on that link. 

------------


> Replacing the build ID with hash is totally 👍 to me,

awesome! that's great news.

Now that you've given an inch, i'll take a mile and ask for the short SHA. :)  


---------

so in total, i'm proposing going from:

https://lhci-canary.herokuapp.com/app/projects/d1e4b15c-e644-4552-b136-e975f486a2ce/builds/39f3f382-a7be-4792-b028-550864806091

to:

https://lhci-canary.herokuapp.com/app/web.dev/compare/55aa605

**Comment by patrickhulce** [comment]
Gotcha, thanks! I get and agree that folks should feel comfortable clicking on these links 👍 

My only issue with `https://lhci-canary.herokuapp.com/app/web.dev/compare/55aa605` is it is unlikely to be globally unique in our immediate next iteration (and not guaranteed to be unique right now either), so it takes the process of creating each record from a simple uuid call to a new bespoke "try to find unique slug, check if it already exists, add randomness, repeat until unique" process. IMO, that tradeoff definitely wasn't worth it for shorter URLs at this point yet but it sounds like it is to you. It would also be a breaking-ish change when we switch so it's easier for our users to just do it now 👍
